servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/ConnectionObserver.java

```diff
@@ -144,9 +144,22 @@
     interface StreamObserver {
 
         /**
-         * Callback when the stream is established and ready.
+         * Callback when a {@code streamId} is assigned.
+         * <p>
+         * Stream identifier may be deferred until after the first write is made on a newly established stream.
+         *
+         * @param streamId assigned stream identifier
+         */
+        default void streamIdAssigned(long streamId) {  // Use long to comply with HTTP/3 requirements
```

Comment:
for visibility and to confirm I understand the rational of this change ... if `StreamObserver` has access to some object representing the stream they could use `stream.getId()` instead of forcing the `long` parameter here, and they could also use this elsewhere. however the thought is since we don't have a good place to put such an abstraction on the base APIs enhancing the observer in this way is the lowest impact change?

Reply:
> since we don't have a good place to put such an abstraction on the base APIs enhancing the observer in this way is the lowest impact change?

Correct.

The initial thought was to make `StreamObserver#streamEstablished()` similar to `ConnectionObserver#connectionEstablished(ConnectionInfo)` and give it some `StreamInfo` object with `#id()` method. However, on the client-side, the `streamEstablished()` is invoked before the first write => stream id is not assigned yet. We discussed this in the past and concluded that giving uses something like `StreamInfo#id()` that can return a different id if invoked before/after the first write is not convenient and complicates understanding.

Instead, this PR introduces a new callback, which is invoked exactly at the time when the id is assigned for a stream. It better represents what is happening at the protocol level.